store/postgres/src/relational.rs

```diff
@@ -1331,6 +1338,130 @@ fn is_object_type(field_type: &q::Type, enums: &EnumMap) -> bool {
+/// Refreshing happens one at a time, and the cache makes sure we minimize
+/// blocking while a refresh happens, favoring using an expired layout over
+/// a refreshed one.
+pub struct LayoutCache {
+    entries: Mutex<HashMap<DeploymentHash, CacheEntry>>,
+    ttl: Duration,
+    /// Use this so that we only refresh one layout at any given time to
+    /// avoid refreshing the same layout multiple times
+    refresh: Mutex<()>,
+}
+
+impl LayoutCache {
+    pub fn new(ttl: Duration) -> Self {
+        Self {
+            entries: Mutex::new(HashMap::new()),
+            ttl,
+            refresh: Mutex::new(()),
+        }
+    }
+
+    fn load(conn: &PgConnection, site: Arc<Site>) -> Result<Arc<Layout>, StoreError> {
+        let subgraph_schema = deployment::schema(conn, site.as_ref())?;
+        let has_poi = crate::catalog::supports_proof_of_indexing(conn, &site.namespace)?;
+        let catalog = Catalog::new(conn, site.clone())?;
+        let layout = Arc::new(Layout::new(
+            site.clone(),
+            &subgraph_schema,
+            catalog,
+            has_poi,
+        )?);
+        Ok(layout)
+    }
+
+    fn cache(&self, layout: Arc<Layout>) {
+        if layout.is_cacheable() {
+            let deployment = layout.site.deployment.clone();
+            let entry = CacheEntry {
+                expires: Instant::now() + self.ttl,
+                value: layout,
+            };
+            &self.entries.lock().unwrap().insert(deployment, entry);
+        }
+    }
+
+    /// Return the corresponding layout if we have one in cache already, and
+    /// ignore expiration infromation
+    pub(crate) fn find(&self, site: &Site) -> Option<Arc<Layout>> {
+        self.entries
+            .lock()
+            .unwrap()
+            .get(&site.deployment)
+            .map(|CacheEntry { value, expires: _ }| value.clone())
+    }
+
+    /// Get the layout for `site`. If it's not in cache, load it. If it is
+    /// expired, try to refresh it if there isn't another refresh happening
+    /// already
+    pub fn get(
+        &self,
+        logger: &Logger,
+        conn: &PgConnection,
+        site: Arc<Site>,
+    ) -> Result<Arc<Layout>, StoreError> {
+        let now = Instant::now();
+        let entry = {
+            let lock = self.entries.lock().unwrap();
+            lock.get(&site.deployment).cloned()
+        };
+        match entry {
+            Some(CacheEntry { value, expires }) => {
+                if now <= expires {
+                    // Entry is not expired; use it
+                    return Ok(value);
+                } else {
+                    // Only do a cache refresh once; we don't want to have
+                    // multiple threads refreshing the same layout
+                    // simultaneously. It's easiest to refresh at most one
+                    // layout globally
+                    let refresh = self.refresh.try_lock();
+                    if let Err(_) = refresh {
```

Comment:
I usually just unwrap poison errors. Or use the `parking_lot` mutex which doesn't poison.

Reply:
This isn't about poisoning, but about whether we would block if we tried to lock, in which case we just return the old layout

Reply:
oh of course, that's what happens when I try to do reviews at the end of the day.